Add addon-test-support tree.

rwjblue opened the pull request:
This commit adds a new tree type to be used during the build processing.  It adds a way for addons to provide testing helpers (in `<addon-name>/test-support` to the consuming app) in the `test-support` file (essentially "vendor" for "tests").

---

- [x] Sanity check
- [x] Add tests
- [x] Perf tests before / after

stefanpenner commented on lib/broccoli/ember-app.js:
```diff
@@ -1241,23 +1241,32 @@ EmberApp.prototype.testFiles = function(coreTestTree) {
+  });
+
+  var inputFiles = [].concat(
+    this.legacyFilesToAppend,
```
i think this wants to be: `legacyTestFilesToAppend`

rwjblue replied:
WHOOPS! yep, good catch

stefanpenner commented:
@rwjblue if you can, it would be great to just double check this doesn't regress the performance in a real app. Nothing is jumping out, but I would like to have PR's author that change the build itself to do some manual tests comparing before/after build times.

https://github.com/ember-cli/stress-app may be a good candidate, but your own apps are also good.

rwjblue commented:
Currently, it should actually result in a bit faster build when no addons are present that utilize the new tree (because I replaced two back to back concats with one smarter one).

But I definitely agree with you!  I added perf tests to the checklist above.

stefanpenner commented:
> Currently, it should actually result in a bit faster build when no addons are present that utilize the new tree (because I replaced two back to back concats with one smarter one).

Ya i noticed, that is nice. I also suspect it will be better. (It would be great to see confirmation of this)

rwjblue commented:
> It would be great to see confirmation of this

Ya, definitely agreed.  I spiked it out to support the changes in https://github.com/ember-cli/ember-cli-qunit/pull/92, and wanted a quick sanity check before continuing down the rabbit hole :rabbit: :rabbit2:

stefanpenner commented:
Ya it seems good

rwjblue commented:
Times using ember-cli/stress-app:

| type | time before | time after |
--------|-------|---------|
|  Initial Cold Build | 164441ms | 173884ms |
| Initial Warm Build | 12040ms | 10264ms
| Rebuild (JS file change) | 4879ms | 5057ms |
| Rebuild (SCSS file change) | 5713ms | 4510ms |

stefanpenner commented:
> Rebuild (JS file change)	4879ms	5057ms

the baseline numbers don't seem correct (they should be around 2s, unless you are on battery or something), can you include the slow tree?

rwjblue commented:
> can you include the slow tree?

Yep, I kept them in [this document](https://usecanvas.com/rwjblue/addon-test-support-timings/3NAMy5kG2T5zibZ5APpL71).

stefanpenner commented:
@rwjblue those merge trees look out of place, what is the result of `npm ls broccoli-merge-trees` ?

rwjblue commented:
> what is the result of npm ls broccoli-merge-trees ?

https://gist.github.com/rwjblue/1adb26dae4f53451968a

stefanpenner commented:
@rwjblue ah ya, that explains it you 1.0.0 should be 1.1.0 for ember-cli (ideally the rest aswell) it seems like you npm install trolled you. As ember-cli actually depends o ^1.0.0 which should have drifted to https://github.com/ember-cli/ember-cli/blob/master/package.json#L51

rwjblue commented:
Ugh.  OK, will `rm -rf node_modules && npm install` in all the repos and try again :sob:

stefanpenner commented:
ember-intl can be bumped to ^2.0.0 (which should promote all the old merge-trees  i see aswell)

rwjblue commented:
> ember-intl can be bumped to ^2.0.0 

Ya, I spotted that also and PR'ed in https://github.com/ember-cli/stress-app/pull/12.

stefanpenner commented:
last remaining one will be addressed by: https://github.com/jasonmit/ember-intl-messageformat-parser/pull/1

stefanpenner commented:
oh and by: https://github.com/aexmachina/ember-cli-sass/commit/73bdf7963335252b2e7e788ebfd7d75a92604060

stefanpenner commented:
with all that, i see:

```
file changed app.js

Build successful - 2081ms.

Slowest Trees                                 | Total
----------------------------------------------+---------------------
SassCompiler                                  | 1403ms
ConcatWithMaps: Concat: App                   | 189ms

Slowest Trees (cumulative)                    | Total (avg)
----------------------------------------------+---------------------
SassCompiler (1)                              | 1403ms
ConcatWithMaps: Concat: App (1)               | 189ms
Babel (17)                                    | 113ms (6 ms)
```

rwjblue commented:
OK, updated times (sorry for the confusion earlier):

| type | time before | time after |
--------|-------|---------|
| Initial Warm Build | 7951ms | 8894ms
| Rebuild (JS file change) | 3564ms | 2494ms |
| Rebuild (SCSS file change) | 3061ms | 2479ms |

Looks like rebuild speeds improved a tiny bit...

rwjblue commented:
@homu r=stefanpenner

homu commented:
:pushpin: Commit 7a7aefc has been approved by `stefanpenner`

<!-- @homu r=stefanpenner 7a7aefcc49855f75f73d4fae172c12a12112815d -->

stefanpenner commented:
@homu r+

homu commented:
:hourglass: Testing commit 7a7aefc with merge 9f663e4...

homu commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #5205

homu commented:
:pushpin: Commit 7a7aefc has been approved by `stefanpenner`

<!-- @homu r=stefanpenner 7a7aefcc49855f75f73d4fae172c12a12112815d -->

homu commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #5205

stefanpenner commented:
tested without sass (satting the masssive 86,000 files in bower) and the rebuild times are like 600ms or so)  so this is good 2 go

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/ember-cli/ember-cli/builds/96608852)
